Generalized indices: `mpp global field` updates

J-Lentz opened the pull request:
**Description**
This PR adds `mpp_global_field` updates for generalized indices, and causes the unit tests added in #1763 to pass. The `MPP_DO_GLOBAL_FIELD_3D_` subroutine is renamed to `MPP_GLOBAL_FIELD_`, which now takes assumed-rank `local` and `global` arguments. Optional `xdim` and `ydim` arguments have also been added, which can be used to specify which dimensions are domain-decomposed. `MPP_DO_GLOBAL_FIELD_A2A_3D_` is removed, as are the various wrapper subroutines for `mpp_global_field` (`MPP_GLOBAL_FIELD_2D_`, `MPP_GLOBAL_FIELD_4D_`, etc). This PR also adds `mpp_domains_mod`, which contains new subroutines for packing/unpacking 2D-5D arrays.

**How Has This Been Tested?**
Builds on C5 with intel-classic. `mpp_global_field` unit tests pass.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

Commit:
Add unit tests for generalized indices

Refactor the mpp_global_field unit tests to test nonstandard dimension
orders. These new tests do not currently pass, as the mpp_global_field
implementation uses hardcoded dimension orders.

Commit:
Break up long lines

Commit:
mpp_*_group_update tests with generalized indices

Test the `mpp_*_group_update` subroutines with nonstandard dimension orders. A
separate module for shared unit test code has also been created.

Commit:
interpolator generalized indices tests

Modify the interpolator tests to test all possible dimension orders.

Commit:
2D generalized indices tests for data_override

The 2D tests in test_data_override_ongrid.F90 have been modified to test
for generalized indices support.

Commit:
Tentative mpp_global_field updates

Commit:
Update mpp_global_field unit tests

Update mpp_global_field unit tests to pass `xdim` and `ydim` arguments. Fix
issues with the rewritten mpp_global_field that cause build errors and test
failures.

Commit:
Move packing/unpacking routines into new file

Create a new file for `arr2vec`, `vec2arr`, and `arr_init`. Add
interfaces for these new subroutines to mpp_domains_mod.

Commit:
Merge remote-tracking branch 'origin/main' into general_indices_mpp_global_field_impl

Commit:
Split long line

mlee03 commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+      subroutine ARR2VEC_ (arr, vec, ix, iy, is, ie, js, je)
+        MPP_TYPE_, intent(in) :: arr(..) !< The array to be packed
+        MPP_TYPE_, intent(out) :: vec(:) !< The vector to copy the data into
+        integer, intent(in) :: ix, iy !< Indices of the domain-decomposed dimensions
```
Just repeating what was mentioned at yesterday's meeting about ix, iy being unclear.  Maybe the word axes is clearer?  An example would be great

J-Lentz replied:
Changed to `xdim` and `ydim` in f42babd091794d9619f0c4e8b4e0b524069be6ad.

mlee03 commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+        MPP_TYPE_, intent(in) :: arr(..) !< The array to be packed
+        MPP_TYPE_, intent(out) :: vec(:) !< The vector to copy the data into
+        integer, intent(in) :: ix, iy !< Indices of the domain-decomposed dimensions
+        integer, intent(in) :: is, ie, js, je !< Starting and ending indices of the x and y dimensions
```
Likewise, examples.. something like `array(is:ie, js:je)` for `ix=1`, `iy=2`
and `array(js:je, is:ie)` for `ix=2`, `iy=1`?

J-Lentz replied:
I've added some examples to the documentation for `is, ie, js, je`.

Commit:
Merge remote-tracking branch 'origin/main' into general_indices_mpp_global_field_impl

Commit:
Workaround for GFortran bug

Use a temporary variable to store the `type(c_ptr)` when passing the
result of `c_loc()` to `c_f_pointer`.

Commit:
Improve documentation

J-Lentz commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+        integer, allocatable, dimension(:) :: lb, ub
+        integer :: n, m
+        integer :: i1, i2, i3, i4, i5
+
```
Added in f42babd091794d9619f0c4e8b4e0b524069be6ad.

J-Lentz commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+        ub(ix) = ie
+        ub(iy) = je
+
+        m = 0
```
I've added an explanation to the declaration of `lb` and `ub`, which I think explains what this block is doing.

J-Lentz commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+      end subroutine ARR2VEC_
+
+      !> Unpack a vector into an array
+      !> @ingroup mpp_domains_mod
```
Analogous comments added in f42babd091794d9619f0c4e8b4e0b524069be6ad.

J-Lentz commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+
+      !> Initialize an assumed-rank array
+      !> @ingroup mpp_domains_mod
+      subroutine ARR_INIT_ (arr, val)
```
Brief explanation added in f42babd091794d9619f0c4e8b4e0b524069be6ad.

J-Lentz commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,173 @@
+!* governing permissions and limitations under the License.
+!***********************************************************************
+
+      !> Pack an array into a vector
```
I've added two sentence explanations of packing and unpacking in f42babd091794d9619f0c4e8b4e0b524069be6ad.

Commit:
Add target attribute to `mpp_domains_stack` in non-MPI build

Commit:
Attempt to fix CMake build

Commit:
Fix CMake build: Attempt 2

Commit:
Fix CMake build: Attempt 3

Commit:
Use permute_arr in mpp_global_field tests

uramirez8707 commented on CMakeLists.txt:
Why are the changes in CMakeLists necessary?

J-Lentz replied:
They're not necessary. I did some back-and-forth with Gemini trying to determine the cause of a CI build failure and while none of its suggestions ended up being the actual issue, they do seem like genuine improvements, so I kept them in the PR. I can `git revert` CMakeLists if anyone disagrees.

uramirez8707 replied:
I agree they do seem like genuine improvements. @rem1776 any thoughts?

rem1776 replied:
Yeah those all look like good improvements to me, we should keep them in. Mostly just updates for consistency so shouldn't change any behavior.

uramirez8707 reviewed (dismissed):
Minor question, but this looks to me.

rem1776 commented on mpp/include/mpp_pack.fh:
```diff
@@ -0,0 +1,197 @@
+            do i2=lb(2),ub(2)
+              do i1=lb(1),ub(1)
+                m = m + 1
+                vec(m) = arr(i1, i2)
```
Maybe instead of iterating would it be possible to use the intrinsic reshape instead? Something like:
```
vec = RESHAPE(arr, [shape(vec)])
```
For these loops I think it should be equivalent at least in the data, but im not sure about the contiguous-ness.

J-Lentz replied:
I think `reshape` would work here. I'll try it and see.

J-Lentz replied:
`do` loops replaced with `reshape` in 4c3a05e4137019065f043ede4f1f5f0f502a88d9.

rem1776 replied:
Thanks!

Commit:
Merge remote-tracking branch 'origin/main' into general_indices_mpp_global_field_impl

Commit:
Replace `do` loops with `reshape` intrinsic

Replace nested do loops in mpp_pack.fh with calls to Fortran's `reshape`
intrinsic. Additionally, the new `mpp_global_field` tests have been
enabled.
